tests/test_knapsack.py

```diff
@@ -15,50 +15,67 @@
+        
+    def test_small_problem(self):
+        """Test the small.csv problem"""
+        data_file_name = os.path.join(example_dir, 'data/small.csv')
+        self._exact_solver_driver(data_file_name, 50, -205, [4, 5, 6])
 
-        value = subprocess.check_output([sys.executable, file_path,
-                                         data_file_path, test_case_weight])
+    def test_very_small_problem(self):
+        """Test the very_small.csv problem"""
+        data_file_name = os.path.join(example_dir, 'data/very_small.csv')
+        self._exact_solver_driver(data_file_name, 10, -10, [0])
 
-        # Check the expected energy
-        energy_expected = "-205"
-        self.assertTrue(energy_expected in str(value))
 
-        # Extract the list from the solution
-        z = str(value).split(" at energy")
-        soln = z[0].split("Found solution ")
-        soln_expected = [20, 10, 15]
-        self.assertEqual(eval(soln[1]), soln_expected)
+class IntegrationTest(unittest.TestCase):
+    """Test the main program run as a script using the small.csv data"""
 
-    def test_knapsack(self):
-        """ Verify contents of output """
+    @classmethod
+    def setUpClass(cls):
+        """Utility method that runs the problem and stores the output"""
         
-        file_path = os.path.join(example_dir, 'main.py')
+        file_path = os.path.join(example_dir, 'knapsack.py')
         data_file_path = os.path.join(example_dir, 'data/small.csv')
-        output = subprocess.check_output([sys.executable, file_path,
-                                         data_file_path, "50"])
-        output = str(output).upper()
-        if os.getenv('DEBUG_OUTPUT'):
-            print("Example output :-"+ output)
-
-        with self.subTest(msg="Verify if output contains 'FOUND SOLUTION' \n"):
-            self.assertIn("FOUND SOLUTION", output)
-        with self.subTest(msg="Verify if output contains correct energy' \n"):
-            self.assertIn("ENERGY -205.0", output)
-        with self.subTest(msg="Verify if error string contains in output \n"):
-            self.assertNotIn("ERROR", output)
-        with self.subTest(msg="Verify if warning string contains in output \n"):
-            self.assertNotIn("WARNING", output)
+        test_case_weight = '50'
+
+        cls.output = subprocess.check_output([sys.executable, file_path,
+                                              data_file_path, test_case_weight])
+        cls.output = cls.output.decode('utf-8') # Bytes to str
+
+    def test_smoke(self):
+        """Verify that the executable script runs and reports some solution"""
+
+        self.assertIn("found solution", self.output.lower())
+
+    def test_solution(self):
+        """Verify that the expected solution is obtained"""
+
+        energy = int(float(self.output.split('\n')[0].split()[-1].strip()))
+        item_indices = self.output.split('\n')[1].split(':')[1]
+        self.assertEqual(eval(item_indices), [4, 5, 6])
```

Comment:
Generally better/safer to use `ast.literal_eval` instead of `eval` for parsing.

Also, the order or item indices is not guaranteed before py37. I'd test set equality. Or sorted lists.

Reply:
Done.  The list of item indices is now sorted before it gets returned from `solve_knapsack`.